core/src/Streamly/Internal/FileSystem/Path/Common.hs

```diff
@@ -269,63 +280,148 @@ hasDrive a
+-- Location or Segment
+------------------------------------------------------------------------------
+
+-- XXX API for static processing of .. (normalizeParentRefs)
+--
+-- Note: paths starting with . or .. are ambiguous and can be considered
+-- segments or rooted. We consider a path starting with "." as rooted, when
+-- someone uses "./x" they explicitly mean x in the current directory whereas
+-- just "x" can be taken to mean a path segment without any specific root.
+-- However, in typed paths the programmer can convey the meaning whether they
+-- mean it as a segment or a rooted path. So even "./x" can potentially be used
+-- as a segment which can just mean "x".
 --
--- On Windows:
--- * @C:\\@ local absolute
--- * @C:@ local relative
--- * @\\@ local relative to current drive root
--- * @\\\\@ UNC network location
+-- XXX For the untyped Path we can allow appending "./x" to other paths. We can
+-- leave this to the programmer. In typed paths we can allow "./x" in segments.
+--
+-- XXX C:\\ is invalid, \\share\ is invalid?
+-- XXX Empty path can be taken to mean "." except in case of UNC paths
+
+-- | Rooted paths on Posix and Windows,
+-- * @/...@ a path starting with a separator
+-- * @.@ current dir
+-- * @./...@ a location relative to current dir
+--
+-- Rooted paths on Windows:
+-- * @C:@ local drive cur dir location
+-- * @C:a\\b@ local drive relative to cur dir location
+-- * @C:\\@ local drive absolute location
+-- * @\\@ local path relative to current drive
+-- * @\\\\share\\@ UNC network location
 -- * @\\\\?\\C:\\@ Long UNC local path
 -- * @\\\\?\\UNC\\@ Long UNC server location
 -- * @\\\\.\\@ DOS local device namespace
 -- * @\\\\??\\@ DOS global namespace
 --
--- * @C:file@ a path relative to curdir.
 isLocation :: (Unbox a, Integral a) => OS -> Array a -> Bool
 isLocation Posix a =
-    -- Assuming path is not empty.
-    isSeparator Posix (wordToChar (Array.getIndexUnsafe 0 a))
-        || isCurDirRelativeLocation a
+    isRelativeCurDrive Posix a
+        || isRelativeCurDir a
 isLocation Windows a =
-    isLocationDrive a
-        || isCurDriveRelativeLocation a
+    isRelativeCurDrive Windows a
+        || isRelativeCurDir a
         || hasDrive a -- curdir-in-drive relative, drive absolute
-        || isAbsoluteUNCLocation a
-        || isCurDirRelativeLocation a
+        || isAbsoluteUNC a
 
 isSegment :: (Unbox a, Integral a) => OS -> Array a -> Bool
 isSegment os = not . isLocation os
 
+------------------------------------------------------------------------------
+-- File or Dir
+------------------------------------------------------------------------------
+
+-- | Returns () if the path can be a valid file, otherwise throws an
+-- exception.
+maybeFile :: (MonadThrow m, Unbox a, Integral a) => OS -> Array a -> m ()
+maybeFile os arr = do
+    s1 <-
+            Stream.toList
+                $ Stream.take 3
+                $ Stream.takeWhile (not . isSeparator os)
+                $ fmap wordToChar
```

Comment:
We are converting all words to characters.
If the input is UTF16 encoded, will all the Word16 translate into proper characters?
We should instead use `isSeperatorIntegral,` which checks the underlying `Word` instead of converting it into a `Char. '

Reply:
I do not see a real problem here.

Reply:
Ah, yes. The problem might arise when we are printing it. You can ignore this comment.